internal/controller/snapshot/snapshot_adapter.go

```diff
@@ -1058,3 +1072,123 @@ func (a *Adapter) findSnapshotWithOpenedPR(snapshot *applicationapiv1alpha1.Snap
+	for i := 1; i < len(sortedSnapshots); i++ {
+
+		if gitops.IsGroupSnapshot(&sortedSnapshots[i]) {
+			// check for pr-group-sha
+			fmt.Println("Yes its group snapshot")
+			latestGroupSha, _ := gitops.GetPRGroupFromSnapshot(&latestSnapshot)
+			groupSha, _ := gitops.GetPRGroupFromSnapshot(&sortedSnapshots[i])
+			if groupSha == latestGroupSha {
+				// this is the snapshot to which we can cancel pipelinerun
+				// set this snapshot as cancelled, superseded
+				err = gitops.MarkSnapshotAsCanceled(a.context, a.client, &sortedSnapshots[i], "Snapshot canceled/superseded")
+				if err != nil {
+					a.logger.Error(err, "Failed to mark snapshot as canceled", "snapshot.Name", snapshot.Name)
+					return err
+				}
+				// now get all integration pipelineruns for a snapshot
+				integrationTestPipelineruns, err := getAllIntegrationPipelineRunsForSnapshot(a.context, a.client, &sortedSnapshots[i])
+				if err != nil {
+					a.logger.Error(err, "Failed to get all integration pipelineruns for snapshot", "snapshot.Name", snapshot.Name)
+					return err
+				}
+				for _, plr := range integrationTestPipelineruns {
+					plr := plr
+					pipelinerunStatus, _, err := h.GetRunningIntegrationPipelineRunStatus(a.context, a.client, &plr)
+					if err != nil {
+						a.logger.Error(err, "Failed to get integration test pipelinerun status", "snapshot.Name", plr.Name)
+						return err
+					}
+					if pipelinerunStatus == intgteststat.IntegrationTestStatusInProgress {
+						//remove finalizer and cancel pipelinerun
+						err = h.RemoveFinalizerFromPipelineRun(a.context, a.client, a.logger, &plr, h.IntegrationPipelineRunFinalizer)
+						if err != nil {
+							a.logger.Error(err, "Failed to remove finalizer from integration test pipelinerun")
+							return err
+						}
+						// set "CancelledRunFinally" to PLR status, should gracefully cancel pipelinerun, this is so raw I hate this
+						patch := client.MergeFrom(plr.DeepCopy())
+						plr.Spec.Status = tektonv1.PipelineRunSpecStatusCancelledRunFinally
+						err := a.client.Status().Patch(a.context, &plr, patch)
+						if err != nil {
+							return err
+						}
+
+					}
+				}
+			}
+		}
+		if gitops.IsComponentSnapshot(&sortedSnapshots[i]) {
+
+			if snapshot.GetAnnotations()[gitops.PipelineAsCodeRepoURLAnnotation] == sortedSnapshots[i].GetAnnotations()[gitops.PipelineAsCodeRepoURLAnnotation] && snapshot.GetAnnotations()[gitops.PipelineAsCodePullRequestAnnotation] == sortedSnapshots[i].GetAnnotations()[gitops.PipelineAsCodePullRequestAnnotation] {
+				err = gitops.MarkSnapshotAsCanceled(a.context, a.client, &sortedSnapshots[i], "Snapshot canceled/superseded")
+				if err != nil {
+					a.logger.Error(err, "Failed to mark snapshot as canceled", "snapshot.Name", snapshot.Name)
+					return err
+				}
+				// get all integration pipelineruns for a snapshot
+				integrationTestPipelineruns, err := getAllIntegrationPipelineRunsForSnapshot(a.context, a.client, &sortedSnapshots[i])
```

Comment:
I feel you may get the integrationPipelinerunsForPR out of the loop, instead of getting integrationPipelineruns for each snapshot. HDYT?

Reply:
I'm not sure if I follow, we get all integration pipelineruns for a single snapshot, then we check for the ones that are running and cancel them.

Reply:
If I understand correctly, we only keep the latest component snapshot and its integration pipelinerun for the PR, is it?

Reply:
Yes, correct.

Reply:
Then we just need the latest PR snapshot and its latest integration test, others ps snapshot and integration test are not needed.

Reply:
I see the poit, I'm just not sure how else we could cancel the rest of the pipelineruns without getting all of them.

Reply:
I was under asumption that the latest snapshot also contains the latest PLR, thats why I never check for both of those and I cancel all others snapshots/PLRs. You think that current behavior is not following this logic as it should?

Reply:
Added the logic, let me know if that fullfills the requirements.